[RFC] Hide list implementation

ZyX-I opened the pull request:
Huge refactoring aiming at plaing with list implementation (specifically, replacing list with either growarray or the like or kvec and running benchmarks). To do that all implementation-specific code should end up in eval/typval.[ch].

Should solve a bunch of issues from #4615 in process, simply because “NULL is an empty list” implementation detail gets hidden away.

ZyX-I committed:
eval/typval: Add functions useful for hiding list implementation

ZyX-I committed:
eval/typval: Add macros useful for hiding list item implementation

ZyX-I committed:
*: Start hiding list implementation

Most of files, except for eval.c and eval/* were only processed by perl.

ZyX-I committed:
eval: Fix inputlist()

ZyX-I committed:
eval: Fix writefile()

ZyX-I committed:
*: Hide list implementation in other files as well

ZyX-I committed:
eval,functests: Fix tests and complete() and setline() behaviour

ZyX-I committed:
eval: Fix setmatches(), setqflist() and setloclist()

ZyX-I committed:
quickfix: Fix :cexpr and :lexpr

ZyX-I committed:
functests: Mark islocked("v:_null_list") behaviour correct

It is the same for other VAR_FIXED lists.

ZyX-I committed:
eval/encode: Fix crash in json_encode test suite

ZyX-I committed:
eval: Fix uniq() crash in legacy test 055

ZyX-I committed:
functests: Add some more NULL tests

ZyX-I committed:
eval: Still check for NULL when doing :unlet

ZyX-I committed:
*: Fix test failures

ZyX-I committed:
*: Finish hiding list implementation

ZyX-I committed:
*: Fix linter errors

ZyX-I committed:
eval: Fix add()

ZyX-I commented:
Should be ready if CI succeeds. Not switching implementation in this PR yet, better to have first refactoring merged and then do some experimenting as it is going to be more subjective (specifically what benchmarks to perform). Note: after this merged any new code matching `->lv_` or `->li_` is invalid with a few exceptions regarding GC and eval/typval.

justinmk commented:
> Note: after this merged any new code matching `->lv_` or `->li_` is invalid with a few exceptions regarding GC and eval/typval.

Perhaps the linter could catch this?

oni-link commented on src/nvim/eval/typval.h:
```diff
@@ -284,20 +303,175 @@ typedef struct list_stack_S {
+
+/// Normalize index: that is, return either -1 or non-negative index
+///
+/// @param[in]  l  List to intex. Used to get length.
```
intex->index

ZyX-I committed:
eval/typval: Fix typo

[ci skip]

ZyX-I commented:
AppVeyor happens to cancel its build even if new pushed commit has `[ci skip]`. Not good, but it only cancelled `MINGW_64-gcov`, other builds had already succeeded.

ZyX-I commented:
Though it started to build a new commit instead.

oni-link commented on src/nvim/eval/typval.h:
```diff
@@ -352,6 +526,75 @@ extern const char *const tv_empty_string;
+///                   also work if you have any uses for the volatile list.
+/// @param[in]  l  List to iterate over.
+/// @param  li  Name of the variable with current listitem_T entry.
+/// @param  code  Cycle body.
```
Perhaps use loop instead of cycle?

ZyX-I replied:
Why? They are synonims and that was already used for TV_DICT_ITER. And loop is also commonly used as a shortcut for “event loop” which has nothing to do with _ITER macros.

ZyX-I commented:
AppVeyor failure looks unrelated:

```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] C:/projects/neovim/test/functional\core\channels_spec.lua @ 60: channels can use stdio channel
.\test\functional\helpers.lua:110: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'stderr'
 *[3] = {
    [1] = 3
   *[2] = {
'
'
'
'
'
'
      [7] = '' } } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'stderr'
 *[3] = {
    [1] = 3
   *[2] = {
     *[1] = '*dies*' } } }
stack traceback:
	.\test\functional\helpers.lua:110: in function 'expect_twostreams'
	C:/projects/neovim/test/functional\core\channels_spec.lua:92: in function <C:/projects/neovim/test/functional\core\channels_spec.lua:60>
[  FAILED  ] C:/projects/neovim/test/functional\core\startup_spec.lua @ 22: startup pipe at both ends: has("ttyin")==0 has("ttyout")==0
C:/projects/neovim/test/functional\core\startup_spec.lua:28: Expected objects to be the same.
Passed in:
(string) '0 0profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\api\buffer.c.gcda:Data file mismatch - some data files may have been concurrently updated without locking support
profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\if_cscope.c.gcda:Merge mismatch for function 20
profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\ops.c.gcda:Merge mismatch for function 23
profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\regexp.c.gcda:Merge mismatch for function 15
profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\search.c.gcda:Merge mismatch for function 36
profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\shada.c.gcda:Merge mismatch for function 2
'
Expected:
(string) '0 0'
stack traceback:
	C:/projects/neovim/test/functional\core\startup_spec.lua:28: in function <C:/projects/neovim/test/functional\core\startup_spec.lua:22>
```

ZyX-I committed:
Merge branch 'master' into hide-container-impl

ZyX-I committed:
*: Fix linter errors

ZyX-I committed:
clint,eval: Make linter check for direct usage of list attributes

bfredl commented:
In this case channels failure is probably caused by gcov failure; gcov tends to add junk also to child process stderr when there is a problem.

ZyX-I commented:
I have added a linter check, and fixed a few places where it matches.

ZyX-I commented:
I am wondering whether gcov can be told to either write to its own log file or just shut up completely. Can’t find something regarding the thing in Google or its documentation though.

ZyX-I commented:
QB hanged, looks unrelated:

```
00:37:08,679 INFO  - # test/functional/core/job_spec.lua @ 222: jobs disallows jobsend on a job that closed stdin
00:37:08,681 INFO  - ok 343 - jobs disallows jobsend on a job that closed stdin
00:42:54,657 INFO  - Terminating launched command gracefully...
00:42:54,657 DEBUG - Executing command: sh /usr/home/quickbuild/buildagent/framework/configuration/org.eclipse.osgi/com/pmease/quickbuild/execution/kill-tree.sh 31472
00:42:54,657 DEBUG - Command working directory: /usr/home/quickbuild/buildagent/bin
00:42:54,666 WARN  - Vim: Caught deadly signal 'SIGTERM'
```

ZyX-I commented:
AV failure is basically the same problem:

```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] C:/projects/neovim/test/functional\core\channels_spec.lua @ 60: channels can use stdio channel
.\test\functional\helpers.lua:110: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'stderr'
 *[3] = {
    [1] = 3
   *[2] = {
'
'
'
      [4] = '' } } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'stderr'
 *[3] = {
    [1] = 3
   *[2] = {
     *[1] = '*dies*' } } }
stack traceback:
	.\test\functional\helpers.lua:110: in function 'expect_twostreams'
	C:/projects/neovim/test/functional\core\channels_spec.lua:92: in function <C:/projects/neovim/test/functional\core\channels_spec.lua:60>
[  FAILED  ] C:/projects/neovim/test/functional\core\startup_spec.lua @ 22: startup pipe at both ends: has("ttyin")==0 has("ttyout")==0
C:/projects/neovim/test/functional\core\startup_spec.lua:28: Expected objects to be the same.
Passed in:
(string) '0 0profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\api\buffer.c.gcda:Data file mismatch - some data files may have been concurrently updated without locking support
profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\eval\typval.c.gcda:Merge mismatch for function 21
profiling:C:\projects\neovim\build\src\nvim/CMakeFiles\nvim.dir\ex_getln.c.gcda:Merge mismatch for function 99
'
Expected:
(string) '0 0'
stack traceback:
	C:/projects/neovim/test/functional\core\startup_spec.lua:28: in function <C:/projects/neovim/test/functional\core\startup_spec.lua:22>
```

, looks unrelated as well.

ZyX-I commented:
I see travis building dependencies, and not for the first time. Does cache ever work at all?

ZyX-I commented:
Travis failure looks strange, it failed to build luarocks:

```
[ 89%] Performing install step for 'luarocks'
make[4]: Entering directory `/home/travis/build/neovim/neovim/.deps/build/src/luarocks'
/home/travis/build/neovim/neovim/.deps/usr/bin/luajit: ...vim/neovim/.deps/build/src/luarocks/src/luarocks/cfg.lua:78: attempt to index local 'site_config' (a boolean value)
stack traceback:
	...vim/neovim/.deps/build/src/luarocks/src/luarocks/cfg.lua:78: in main chunk
	[C]: in function 'require'
	src/bin/luarocks:4: in main chunk
	[C]: at 0x00404b20
make[4]: *** [run_luarocks] Error 1
```

. Restarting.

ZyX-I commented:
Travis succeeded, though was showing up as not so I restarted flaky builds after successfull lint build.

oni-link commented on src/nvim/eval/typval.c:
```diff
@@ -339,12 +402,12 @@ void tv_list_append_list(list_T *const list, list_T *const itemlist)
+  TV_LIST_ITEM_TV(li)->v_lock = VAR_UNLOCKED;
+  TV_LIST_ITEM_TV(li)->vval.v_list = itemlist;
   tv_list_append(list, li);
   if (itemlist != NULL) {
```
Could be removed, `tv_list_ref()` checks also for `NULL`.

oni-link commented on src/nvim/eval/typval.c:
```diff
@@ -439,33 +502,32 @@ list_T *tv_list_copy(const vimconv_T *const conv, list_T *const orig,
-  listitem_T *item;
-  for (item = orig->lv_first; item != NULL && !got_int;
-       item = item->li_next) {
+  TV_LIST_ITER(orig, item, {
```
Missing `got_int` check. Was this on purpose?

ZyX-I replied:
It appears that I can have got_int check here, but it is not exactly useful:

```
local helpers = require('test.functional.helpers')(after_each)

local command = helpers.command
local meths = helpers.meths
local clear = helpers.clear
local funcs = helpers.funcs
local sleep = helpers.sleep
local wait = helpers.wait
local feed = helpers.feed
local eq = helpers.eq

local screen
local dur
local min_dur = 8
local len = 131072

describe('List support code', function()
  before_each(function()
    clear()
    if dur == nil then
      dur = 0
      while true do
        command(([[
          let rt = reltime()
          let bl = range(%u)
          let dur = reltimestr(reltime(rt))
        ]]):format(len))
        dur = tonumber(meths.get_var('dur'))
        if dur >= min_dur then
          print(('Using len %u, dur %g'):format(len, dur))
          break
        else
          len = len * 2
        end
      end
    end
  end)
  it('allows interrupting copy', function()
    feed(':let t_rt = reltime()<CR>:let t_bl = copy(bl)<CR>')
    sleep(min_dur / 16 * 1000)
    feed('<C-c>')
    wait()
    command('let t_dur = reltimestr(reltime(t_rt))')
    local t_dur = tonumber(meths.get_var('t_dur'))
    if t_dur >= dur / 8 then
      eq(nil, ('Took too long to cancel: %g >= %g'):format(t_dur, dur / 8))
    end
  end)
end)
```

with got_int check added always fails. Manual testing confirms that copy() accually runs *slower*  when cancelled because the only thing it does cancel is an assignment, which ultimately means that big list needs to be freed before continuing.

Test succeeds if I not only add `if (got_int) break;` but also `os_breakcheck()`.

ZyX-I replied:
Though I should say removal was not *on purpose*.

oni-link commented on src/nvim/eval/typval.c:
```diff
@@ -540,13 +602,12 @@ static int list_join_inner(garray_T *const gap, list_T *const l,
 
   // Stringify each item in the list.
-  for (item = l->lv_first; item != NULL && !got_int; item = item->li_next) {
+  TV_LIST_ITER(l, item, {
```
Missing `got_int` check.

oni-link commented on src/nvim/eval/typval.c:
```diff
@@ -632,18 +693,45 @@ bool tv_list_equal(list_T *const l1, list_T *const l2, const bool ic,
-       ; item1 = item1->li_next, item2 = item2->li_next) {
-    if (!tv_equal(&item1->li_tv, &item2->li_tv, ic, recursive)) {
+       ; (item1 = TV_LIST_ITEM_NEXT(l1, item1),
+          item2 = TV_LIST_ITEM_NEXT(n2, item2))) {
```
Wrong list name: `n2` -> `l2`.

oni-link commented on src/nvim/eval/typval.c:
```diff
@@ -772,15 +855,14 @@ long tv_list_idx_of_item(const list_T *const l, const listitem_T *const item)
-  }
-  return idx;
+  });
+  return -1;
```
The function return value could be changed to `int`.

ZyX-I replied:
Not now. I have no idea why original functions have `long` type (AFAIR I got that from Vim), but that will for sure raise a bunch of warnings.

oni-link commented on src/nvim/eval/typval.c:
```diff
@@ -2027,7 +2109,7 @@ void tv_copy(typval_T *const from, typval_T *const to)
     }
     case VAR_LIST: {
       if (from->vval.v_list != NULL) {
```
Check could be removed.

oni-link commented on src/nvim/quickfix.c:
```diff
@@ -4399,7 +4403,7 @@ void ex_cexpr(exarg_T *eap)
   if (eval0(eap->arg, &tv, NULL, true) != FAIL) {
     if ((tv.v_type == VAR_STRING && tv.vval.v_string != NULL)
-        || (tv.v_type == VAR_LIST && tv.vval.v_list != NULL)) {
+        || tv.v_type == VAR_LIST) {
```
Does `eval0()` returns a list that is not `NULL`?

ZyX-I replied:
? It could return either list, removal of this check makes it sure that NULL list is processed like empty one.

oni-link commented on src/nvim/window.c:
```diff
@@ -5640,28 +5635,28 @@ int match_add(win_T *wp, const char *const grp, const char *const pat,
           continue;
```
The iterative statement of the loop does not increase `i` anymore, so decreasing `i` before a `continue` is wrong.

oni-link commented on src/nvim/window.c:
```diff
@@ -5640,28 +5635,28 @@ int match_add(win_T *wp, const char *const grp, const char *const pat,
-          --i;
+      } else if (TV_LIST_ITEM_TV(li)->v_type == VAR_NUMBER) {
+        if (TV_LIST_ITEM_TV(li)->vval.v_number == 0) {
+          i--;
```
Should also be removed, see other continue.

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -135,21 +136,22 @@ static int conv_error(const char *const msg, const MPConvStack *const mpstack,
-          idx++;
-        }
+        const listitem_T *const li = (v.data.l.li == NULL
+                                      ? tv_list_last(v.data.l.list)
```
In this case we could compute the value of `idx` directly (`MAX(tv_list_len(v.data.l.list)-1,0)`).

oni-link commented on src/nvim/eval/encode.c:
```diff
@@ -202,21 +204,17 @@ bool encode_vim_list_to_buf(const list_T *const list, size_t *const ret_len,
-    if (len) {
-      len--;
+    len++;
+    if (TV_LIST_ITEM_TV(li)->vval.v_string != 0) {
```
`!= 0` -> `!= NULL`

ZyX-I commented:
I am wondering WTF sleep is doing and why is it using `session:…` at all. The following code does not work as expected if I replace `os.execute('sleep …')` with `sleep` from helpers:

```
  it('allows interrupting copy', function()
    feed(':let t_rt = reltime()<CR>:let t_bl = copy(bl)<CR>')
    os.execute('sleep 0.01s')
    -- sleep(min_dur / 16)
    feed('<C-c>')
    wait()
    command('let t_dur = reltimestr(reltime(t_rt))')
    local t_dur = tonumber(meths.get_var('t_dur'))
    eq(true, t_dur < dur / 8)
  end)
```

Specifically by “does not work as expected” I mean “happens to interrupt code before t_rt has a chance to be assigned a value”, no matter which value I provide to sleep exactly.

ZyX-I committed:
*: Fix some problems found during review

Still missing: problems in window.c, it should be possible to construct a test
for them.

ZyX-I committed:
window: Fix matchaddpos() and enhance error reporting

ZyX-I committed:
eval/encode: Avoid unnecessary tv_list_idx_of_item() calls

ZyX-I committed:
eval: Fix linter error

ZyX-I committed:
Merge branch 'master' into hide-container-impl

ZyX-I commented:
If CI passes this should be ready, previous CI only failed trailing whitespace lint check.

ZyX-I commented:
AppVeyor and QB succeeded.

ZyX-I commented:
Travis failure looks unrelated: 

```
missing:
provider/pythonx: Could not load Python 3:
/opt/pyenv/shims/python3 does not have the "neovim" module. :help provider-python
python3.7 not found in search path or not executable.
/opt/pyenv/shims/python3.6 does not exist: pyenv: python3.6: command not found

The `python3.6' command exists in these Python versions:
  3.6
  3.6.3


python3.5 not found in search path or not executable.
/usr/bin/python3.4 does not have the "neovim" module. :help provider-python
python3.3 not found in search path or not executable.
/opt/pyenv/shims/python is Python 2.7 and cannot provide Python 3.[0m
[1m[31m[  ERROR   ][0m[0m [1m1[0m error, listed below:
[1m[31m[  ERROR   ][0m[0m [36m...uild/neovim/neovim/test/functional/terminal/tui_spec.lua[0m @ [36m351[0m: [1mtui FocusGained/FocusLost in cmdline-mode[0m
./test/functional/helpers.lua:269: 
retry() attempts: 2
./test/functional/ui/screen.lua:307: Row 2 did not match.
Expected:
  |{1: }                                                 |
  |*lost                                              |
  |gained                                            |
  |{4:~                                                 }|
  |{5:[No Name] [+]                                     }|
  |:                                                 |
  |{3:-- TERMINAL --}                                    |
Actual:
  |{1: }                                                 |
  |*gained                                            |
  |lost                                              |
  |gained                                            |
  |{5:[No Name] [+]                                     }|
  |:                                                 |
  |{3:-- TERMINAL --}                                    |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.  

stack traceback:
	./test/functional/helpers.lua:269: in function 'retry'
	...uild/neovim/neovim/test/functional/terminal/tui_spec.lua:358: in function <...uild/neovim/neovim/test/functional/terminal/tui_spec.lua:351>


 [1m11[0m SKIPPED TESTS
 [1m1[0m ERROR
-- Output to stderr:

CMake Error at /home/travis/build/neovim/neovim/cmake/RunTests.cmake:53 (message):
  Running functional tests failed with error: 1.
```

. Restarting.

ZyX-I committed:
functests: Fix linter error

ZyX-I commented:
CI succeeded.

justinmk commented:
@oni-link done with comments?

oni-link commented:
@justinmk, I have not yet checked `eval.c`, otherwise done with `.c` files.

oni-link commented on src/nvim/eval.c:
```diff
@@ -5330,10 +5340,13 @@ bool set_ref_in_list(list_T *l, int copyID, ht_stack_T **ht_stack)
+      TV_LIST_ITER(cur_l, li, {
+        abort = set_ref_in_item(TV_LIST_ITEM_TV(li), copyID, ht_stack,
+                                &list_stack);
+        if (abort) {
```
The outer enclosing `if (!abort)` could be removed.

oni-link commented on src/nvim/eval.c:
```diff
@@ -7934,34 +7944,51 @@ static void f_empty(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+          n = false;
+          break;
+        }
+        case kSpecialVarFalse:
```
Do we need a a "fallthrough" comment?

ZyX-I replied:
Where and why? I have all proper breaks everywhere needed here.

oni-link commented on src/nvim/eval.c:
```diff
@@ -8936,7 +8968,7 @@ static void common_function(typval_T *argvars, typval_T *rettv,
         }
         list = argvars[arg_idx].vval.v_list;
-        if (list == NULL || list->lv_len == 0) {
+        if (list == NULL || tv_list_len(list) == 0) {
```
`list == NULL` could be removed.

oni-link commented on src/nvim/eval.c:
```diff
@@ -11025,39 +11055,40 @@ static void f_indent(typval_T *argvars, typval_T *rettv, FunPtr fptr)
-        ic = tv_get_number_chk(&argvars[3], &error);
-      }
-      if (error) {
+      idx = tv_list_uidx(l, tv_get_number_chk(&argvars[2], &error));
```
Why not only error check `tv_get_number_chk()` here and let `tv_list_find()` handle the normalization and error handling?

ZyX-I replied:
And how I am to get the number to return exactly? `tv_list_find()` does not return one and I was not feeling like refactoring it to have ret_idx argument as it is useful in two places out of 18 tv_list_find is being used.

oni-link commented on src/nvim/eval.c:
```diff
@@ -11025,39 +11055,40 @@ static void f_indent(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      }
+      if (argvars[3].v_type != VAR_UNKNOWN) {
+        ic = !!tv_get_number_chk(&argvars[3], &error);
       }
```
In the old code the following loop would not be entered, if `error` was set. `error` should also be set to `false` before the call to `tv_get_number_chk()`  if it should be evaluated.

oni-link commented on src/nvim/eval.c:
```diff
@@ -12221,11 +12247,11 @@ static void find_some_match(typval_T *argvars, typval_T *rettv, int type)
     if (l != NULL) {
-      li = tv_list_find(l, start);
+      idx = tv_list_uidx(l, start);
+      li = tv_list_find(l, idx);
```
Why precompute `idx` if it is recomputed in this function?

oni-link replied:
Ah, `idx` could be used later on. If this is the case computation of `idx` could be moved after the `if` block.

oni-link commented on src/nvim/eval.c:
```diff
@@ -15432,34 +15440,25 @@ static void do_sort_uniq(typval_T *argvars, typval_T *rettv, bool sort)
-          tv_list_item_free(li);
-          l->lv_len--;
+          tv_list_item_remove(l, li);
+          li = tv_list_find(l, idx);
```
Finding the new `li` looks kind of expansive. Can't `prev_li` be reused to set `li`?

ZyX-I replied:
No, it will actually be invalid after removal. Not invalid *with the current implementation*, but invalid with array-based list implementation that may trigger realloc() at removal. I know that this is suboptimal, but I also know that `uniq()` is almost never used.

ZyX-I replied:
After searching for uniq() calls in the plugins that happened to be in my ~/.vam I found exactly *one* user of actual uniq() (that is, neosnippet) and a bunch of uniq() implementations based on dictionary (those that do not do sorting), though I do not quite understand WTF author was thinking when writing such code as

```
function! deoplete#util#uniq(list) abort
  let list = map(copy(a:list), '[v:val, v:val]')
  let i = 0
  let seen = {}
  while i < len(list)
    let key = string(list[i][1])
    if has_key(seen, key)
      call remove(list, i)
    else
      let seen[key] = 1
      let i += 1
    endif
  endwhile
  return map(list, 'v:val[0]')
endfunction
```

(it is not deoplete, same thing is seen in some other plugins). (Basically, what I do not understand is what the first and last map()s are for, from my POV you can just copy list over and do remove().)

ZyX-I replied:
For reference, my ~/.vam contains something like 200 plugins, though most normally not updated as I only installed them to check something.

ZyX-I replied:
Oh, I remembered that I actually bothered to clone all repositories from https://github.com/vim-scripts once in Jun 18, may search that for uniq(). The following is all what `ag` managed to find there:

```
% ag '^\s*(?!")\S.*(?<![#a-zA-Z0-9.]|\b\w:)\buniq\(' -G '.*\.vim' ../vim-scripts | tee uniqs.lst
../vim-scripts/AmbiCompletion.git/plugin/ambicompletion.vim:151:            call uniq(candidates)
../vim-scripts/AmbiCompletion.git/plugin/ambicompletion.vim:316:        call uniq(cachewords)
../vim-scripts/AmbiCompletion.git/plugin/ambicompletion.vim:355:        call uniq(cachewords)
../vim-scripts/changesPlugin.git/autoload/changes.vim:783:          call uniq(b:diffhl[i])
../vim-scripts/changesPlugin.git/autoload/changes.vim:1283:     let lines = uniq(lines)
../vim-scripts/changesPlugin.git/autoload/changes.vim:1347:         let g:lines=uniq(g:lines)
../vim-scripts/clickable.vim.git/autoload/vital/_clickable_vim.vim:265:    return uniq(a:list)
../vim-scripts/cmake.vim--Alcin.git/autoload/cmake/extension.vim:49:
../vim-scripts/cmake.vim--Alcin.git/autoload/cmake/extension.vim:50:  call uniq(l:signatures)
../vim-scripts/cmake.vim--Alcin.git/autoload/cmake/extension/ninja.vim:47:  call uniq(l:flags)
../vim-scripts/cmake.vim--Alcin.git/autoload/cmake/extension/ninja.vim:117:  call uniq(l:ninja_file_lines)
../vim-scripts/columnmove.git/autoload/vital/_columnmove.vim:265:    return uniq(a:list)
../vim-scripts/ctrlp-locate.git/autoload/vital/_ctrlp_locate.vim:276:    return uniq(a:list)
../vim-scripts/Dutyl.git/autoload/dutyl/util.vim:81:        return uniq(sort(a:list))
../vim-scripts/incsearch.vim.git/autoload/vital/_incsearch.vim:282:    return uniq(a:list)
../vim-scripts/ingo-library.git/autoload/ingo/compat.vim:93:    return uniq(a:list)
../vim-scripts/open-browser.vim.git/autoload/vital/_openbrowser.vim:265:    return uniq(a:list)
../vim-scripts/projectionist.vim.git/autoload/projectionist.vim:29:    return uniq(a:list)
../vim-scripts/tsuquyomi.git/autoload/vital/_tsuquyomi.vim:266:    return uniq(a:list)
../vim-scripts/unite-spell-suggest.git/autoload/mklib/string.vim:142:  return a:0 && a:1 ? uniq(sort(l:parts)) : l:parts
../vim-scripts/vim-gista.git/autoload/vital/_vim_gista.vim:247:    return uniq(a:list)
../vim-scripts/vim-javacomplete2.git/autoload/javacomplete/server.vim:15:    return uniq(a:list)
../vim-scripts/vim-operator-flashy.git/autoload/vital/_operator_flashy.vim:272:    return uniq(a:list)
../vim-scripts/vim-pandoc-bundle.git/autoload/pandoc/command.vim:72:            return filter(uniq(extend(sort(short_opts), sort(long_opts))), 'v:val != "-:"')
../vim-scripts/vim-signjk-motion.git/autoload/vital/_signjk.vim:272:    return uniq(a:list)
../vim-scripts/vim-snoopy.git/autoload/vital/_snoopy.vim:266:    return uniq(a:list)
../vim-scripts/vison.git/autoload/vital/_vison.vim:266:    return uniq(a:list)
```

ZyX-I committed:
eval: Fix some issues found in review

ZyX-I commented:
CI succeeded.

justinmk commented:
Conflict in quickfix.c

ZyX-I committed:
Merge branch 'master' into hide-container-impl

ZyX-I commented:
Conflict resolved, though it was almost about nothing: static initializer of the structure I edited was replaced with memset. I thought lists are used more often to cause conflicts with actual list changes.

ZyX-I committed:
regexp: Fix linter errors

ZyX-I committed:
clint: Make linter report line where it found opening brace

justinmk commented on test/functional/eval/interrupt_spec.lua:
```diff
@@ -0,0 +1,61 @@
+local len = 131072
+
+describe('List support code', function()
+  if not pending('does not actually allows interrupting with just got_int', function() end) then return end
```
The `if not pending` check seems like a mistake. Would be clearer:

```
pending('does not actually allows interrupting with just got_int', function() end)
return
```

Or remove it totally?

ZyX-I replied:
I wrote it this way so that it is clear that return is there due to pending and to write everything in one line (one-line if is sometimes used, two syntactically unconnected statements on one line are not).
